src/fcc_bridge_node_ros.cpp

```diff
@@ -0,0 +1,225 @@
+                     "Repeated try to setup ros components");
+        this->internal_state = INTERNAL_STATE::ERROR;
+        return;
+    }
+
+    // Setup publisher
+
+    // Create GPSInfo & Position publisher
+    this->gps_position_publisher =
+        this->create_publisher<interfaces::msg::GPSPosition>("uav_gps_position",
+                                                             1);
+    // Create FlightState publisher
+    this->flight_state_publisher =
+        this->create_publisher<interfaces::msg::FlightState>("uav_flight_state",
+                                                             1);
+
+    // Create BatteryState publisher
+    this->battery_state_publisher =
+        this->create_publisher<interfaces::msg::BatteryState>(
+            "uav_battery_state", 1);
+
+    // Create RCState publisher
+    this->rc_state_publisher =
+        this->create_publisher<interfaces::msg::RCState>("uav_rc_state", 1);
+
+    // Setup subscriber
+
+    // Create subscription options for heartbeat to only receive mission control
+    // heartbeats
+    rclcpp::SubscriptionOptions subscription_options;
+    subscription_options.content_filter_options.filter_expression =
+        "sender_id = 'mission_control'";
+
+    // Create Heartbeat subscription
+    this->mission_control_heartbeat_subscriber =
+        this->create_subscription<interfaces::msg::Heartbeat>(
+            "heartbeat", 1,
+            std::bind(&FCCBridgeNode::mission_control_heartbeat_subscriber_cb,
+                      this, std::placeholders::_1),
+            subscription_options);
+
+    // Ensure that the filter is enabled. (Some DDS versions do not support it)
+    if (!this->mission_control_heartbeat_subscriber->is_cft_enabled()) {
+        RCLCPP_FATAL(this->get_logger(), "Content filtering is not enabled!");
+        this->internal_state = INTERNAL_STATE::ERROR;
+        return;
+    }
+
+    // Setup 5Hz timer to get telemetry from the FCC
+    this->fcc_telemetry_timer_5hz = this->create_wall_timer(
+        std::chrono::milliseconds{200},
+        std::bind(&FCCBridgeNode::fcc_telemetry_timer_5hz_cb, this));
+
+    // Setup 10Hz timer to get telemetry from the FCC
+    this->fcc_telemetry_timer_5hz = this->create_wall_timer(
+        std::chrono::milliseconds{100},
+        std::bind(&FCCBridgeNode::fcc_telemetry_timer_10hz_cb, this));
+}
+
+void FCCBridgeNode::mission_control_heartbeat_subscriber_cb(
+    const interfaces::msg::Heartbeat &msg) {
+    RCLCPP_DEBUG(this->get_logger(), "Received heartbeat from mission control");
+
+    // Check if the heartbeat tick has increased since last time
+    if (msg.tick <= this->last_mission_control_heartbeat.tick) {
+        RCLCPP_ERROR(this->get_logger(),
+                     "The received heartbeat is not newer than the last one! "
+                     "Triggering RTH...");
+        this->trigger_rth();
+    }
+
+    // Check if mission control is still active
+    if (!msg.active) {

```

Comment:
You're getting something wrong here.
The definition of the active flag is:
*You have to set this to true if your node is actively sending data to the interface, and false if not.*

This means that mission control will only be active as long as it is actively sending flight actions to the bridge.
This flag is only there to protect the fcc bridge to receive messages from two nodes at the same time, because something went wrong and both of them are active, even though one of them shouldn't.

The only exception is the fcc bridge itself, because for this the definition doesn't really make sense so where using it as an indicator when the bridge can accept messages from the nodes.

Reply:
Alright then I will completely ignore this value.
But it would then make sense to check if the cached message indicated an active state if I get a take off command. And the inverse for waypoint commands. Right?

Reply:
This is a bit more complex.
You have to track the currently active node by listening on the control topic (docs: https://docs.google.com/document/d/1BV6CUhAO0_3A8xje5DsFQydK13K3FuRfkojx6XGCMnk/edit#heading=h.n1toqlb5yeth).

Then you have to check that only the currently active node is sending messages to you.

But you don't have to care about the active flag in the heartbeat. This is done by mission control.

Reply:
Good to know. I will remove this check here. Subscribe to the active node topic and then do a check when I receive a command.

Reply:
Removed the check in 476c66c

Reply:
The rest will follow once I get to it.